src/ctap/credential_management.rs

```diff
@@ -0,0 +1,884 @@
+    let PublicKeyCredentialSource {
+        key_type,
+        credential_id,
+        private_key,
+        rp_id: _,
+        user_handle,
+        user_display_name,
+        cred_protect_policy,
+        creation_order: _,
+        user_name,
+        user_icon,
+    } = credential;
+    let user = PublicKeyCredentialUserEntity {
+        user_id: user_handle,
+        user_name,
+        user_display_name,
+        user_icon,
+    };
+    let credential_id = PublicKeyCredentialDescriptor {
+        key_type,
+        key_id: credential_id,
+        transports: None, // You can set USB as a hint here.
+    };
+    let public_key = CoseKey::from(private_key.genpk());
+    Ok(AuthenticatorCredentialManagementResponse {
+        existing_resident_credentials_count: None,
+        max_possible_remaining_resident_credentials_count: None,
+        rp: None,
+        rp_id_hash: None,
+        total_rps: None,
+        user: Some(user),
+        credential_id: Some(credential_id),
+        public_key: Some(public_key),
+        total_credentials,
+        cred_protect: cred_protect_policy,
+        // TODO(kaczmarczyck) add when largeBlobKey is implemented
+        large_blob_key: None,
+    })
+}
+
+/// Processes the subcommand getCredsMetadata for CredentialManagement.
+fn process_get_creds_metadata(
+    persistent_store: &PersistentStore,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    Ok(AuthenticatorCredentialManagementResponse {
+        existing_resident_credentials_count: Some(persistent_store.count_credentials()? as u64),
+        max_possible_remaining_resident_credentials_count: Some(
+            persistent_store.remaining_credentials()? as u64,
+        ),
+        rp: None,
+        rp_id_hash: None,
+        total_rps: None,
+        user: None,
+        credential_id: None,
+        public_key: None,
+        total_credentials: None,
+        cred_protect: None,
+        large_blob_key: None,
+    })
+}
+
+/// Processes the subcommand enumerateRPsBegin for CredentialManagement.
+fn process_enumerate_rps_begin(
+    persistent_store: &PersistentStore,
+    stateful_command_permission: &mut TimedPermission,
+    stateful_command_type: &mut Option<StatefulCommand>,
+    now: ClockValue,
+) -> Result<AuthenticatorCredentialManagementResponse, Ctap2StatusCode> {
+    let mut rp_set = BTreeSet::new();
+    let mut iter_result = Ok(());
+    for (_, credential) in persistent_store.iter_credentials(&mut iter_result)? {
+        rp_set.insert(credential.rp_id);
+    }
+    iter_result?;
+    let mut rp_ids = Vec::from_iter(rp_set);
+    let total_rps = rp_ids.len();
+
+    // TODO(kaczmarczyck) behaviour with empty list?
+    let rp_id = rp_ids.pop();
+    if total_rps > 1 {
+        *stateful_command_permission =
+            TimedPermission::granted(now, STATEFUL_COMMAND_TIMEOUT_DURATION);
+        *stateful_command_type = Some(StatefulCommand::EnumerateRps(rp_ids));
```

Comment:
It's a bit hard for me to be sure here...
When we receive an enumeration, we'll generate a `Vec<string>` (which can consume quite a lot of RAM). But even if the permission expires, the object will not be destroyed until the next time we receive a command that will overwrite this, correct?

Reply:
Yes, the main loop logic is very lightweight. The timing of destruction shouldn't change worst case memory consumption IMO. If you are worried about the `Vec<String>`, we could instead keep a storage key of `usize`, iterate the credential keys, keep a set of relying parties that were already used and respond with the next unseen RP ID. @ia0 this should work? Something like the following and an initial counter for how many RP IDs exist.

```rust
self.first_unseen_key = key::CREDENTIAL.start;

fn next_rp() {
  let mut seen_rp_ids = BTreeSet::new();
  for key in key::CREDENTIAL.start..self.first_unseen_key {
    seen_rps.add(rp_id_of_credential_at_key);
  }
  for key in self.first_unseen_key..key::CREDENTIAL.end {
    if seen_rps.contains(rp_id_of_credential_at_key) {
      break;
    }
    self.first_unseen_key += 1;
  }
  // Use RP ID at self.first_unseen_key, increment again, ...
}
```

Reply:
It's not the best to iterate on keys, because looking up a key is currently linear (would be constant with #238). There would be solutions to resume `iter_credentials`, but there might be some lifetime issues. I think the best in this case would simply be to recompute the set of RPids at each command (initial command and each follow-up) since we have a new time budget for each command and we know it's enough because we anyway need to do it once at some point. So the state could simply be the index in the ordered set of the next RP id. What do you think?

Reply:
That works too, as a way to reduce the state we need to keep to `usize`. Thanks for the input. @jmichelp would that help with your concern?

Reply:
It should be fine provided we can't get dirty-read cases: e.g. we have an iterator, then we do some modification on the list (i.e. delete and entry) then we continue iterating.
But I don't think it's very different from this point of view from the current code as we would keep in `stateful_command_type` a list that may point to deleted entries.

Reply:
Done as proposed by @ia0 . I added a test to make sure the sorting is the same each time.